### PR TITLE
jenkins: Add template to enable build Pipelines

### PR DESCRIPTION
**Pull request by aramalipoor**
Docs at https://github.com/abarcloud/docs/pull/113

# Release
```sh
make -e NAMESPACE=openshift jenkins/
```

### CODE REVIEW

File: jenkins/jenkins.yml [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    iconClass: icon-jenkins
+    tags: instant-app,jenkins
+
+message: |-
+    Jenkins service has been created in your project.
+
+      Username: admin
+      Password: ${JENKINS_PASSWORD}
+
+    After a few minutes you will able to log into Jenkins web interface.
```
Comment (alikhajeh1) [inline jenkins/jenkins.yml]: Would be good to put the URL here

File: jenkins/jenkins.yml [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+objects:
+
+- apiVersion: v1
+  kind: Route
+  metadata:
+    annotations:
+      haproxy.router.openshift.io/timeout: 4m
+      template.openshift.io/expose-uri: http://{.spec.host}{.spec.path}
```
Comment (alikhajeh1) [inline jenkins/jenkins.yml]: What is this set to? we auto-generate a .abar.cloud one for pgweb/nuget but give users the option to over-ride it

File: jenkins/jenkins.yml [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+      nodePort: 0
+      port: 8080
+      protocol: TCP
+      targetPort: 8080
+    selector:
+      name: jenkins
+    sessionAffinity: None
+    type: ClusterIP
+
+parameters:
```
Comment (alikhajeh1) [inline jenkins/jenkins.yml]: Other templates use a SERVICE_NAME concept to prefix all resources, so users can deploy multiple instances of templates - helps a lot for testing or when they mess-up

File: jenkins/jenkins.yml [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      Password: ${JENKINS_PASSWORD}
+
+    After a few minutes you will able to log into Jenkins web interface.
+objects:
+
+- apiVersion: v1
+  kind: Route
+  metadata:
+    annotations:
+      haproxy.router.openshift.io/timeout: 4m
```
Comment (alikhajeh1) [inline jenkins/jenkins.yml]: Why is this needed? Jenkins is slow on 1000Mi?

File: jenkins/jenkins.yml [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+      metadata:
+        labels:
+          name: jenkins
+      spec:
+        containers:
+        - env:
+          - name: JENKINS_PASSWORD
+            value: ${JENKINS_PASSWORD}
+          - name: INSTALL_PLUGINS
+            value: "pipeline-githubnotify-step:1.0.4,github-branch-source:2.3.3,github-scm-trait-notification-context:1.1"
```
Comment (alikhajeh1) [inline jenkins/jenkins.yml]: are there gitlab plugins that we can install by the default too? gitlab's popular in iran...

File: jenkins/jenkins.yml [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+kind: Template
+metadata:
+  name: jenkins
+  annotations:
+    openshift.io/display-name: Jenkins
+    openshift.io/support-url: https://docs.abarcloud.com/support.html
+    openshift.io/documentation-url: https://docs.abarcloud.com/learn/continuous-integration.html
+    description: An automation server for the non-human part of the software development process, with continuous integration and facilitating continuous delivery.
+    iconClass: icon-jenkins
+    tags: instant-app,jenkins
```
Comment (alikhajeh1) [inline jenkins/jenkins.yml]: what category does this go in the catalog? Other Services?

File: jenkins/jenkins.yml [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  description: Password for the Jenkins "admin" user. Leave this empty if you'd
+    like it to be auto-generated.
+  from: '[a-zA-Z0-9]{16}'
+  generate: expression
+  required: true
+- description: Maximum amount of memory Jenkins container can use.
+  displayName: Memory Limit
+  name: MEMORY_LIMIT
+  value: 1000Mi
+- description: Volume space available for data, e.g. 512Mi, 2Gi.
```
Comment (alikhajeh1) [inline jenkins/jenkins.yml]: Maybe add a bit to say what data is? build logs? configs? build assets? 
512Mi won't work for storage, so it should be like the other templates (I think 1Gi, 5Gi etc.)

File: jenkins/jenkins.yml [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  displayName: Admin Password
+  description: Password for the Jenkins "admin" user. Leave this empty if you'd
+    like it to be auto-generated.
+  from: '[a-zA-Z0-9]{16}'
+  generate: expression
+  required: true
+- description: Maximum amount of memory Jenkins container can use.
+  displayName: Memory Limit
+  name: MEMORY_LIMIT
+  value: 1000Mi
```
Comment (alikhajeh1) [inline jenkins/jenkins.yml]: This is the minimum Jenkins' works with?

File: jenkins/jenkins.yml [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+  metadata:
+    annotations:
+      template.alpha.openshift.io/wait-for-ready: 'true'
+    name: jenkins
+  spec:
+    replicas: 1
+    selector:
+      name: jenkins
+    strategy:
+      type: Recreate
```
Comment (alikhajeh1) [inline jenkins/jenkins.yml]: Is it possible to have more than replica? If so, this might be better as Rolling? If not, plz add a note in the template desc (see mysql/postgres and use the same wording to say users can't scale beyond one pod).

File: jenkins/jenkins.yml (hunk 4 above)
Comment (aramalipoor) [reply jenkins/jenkins.yml]: Yeah, only on plugins page though. Seems to be looping over every single one of them.

File: jenkins/jenkins.yml (hunk 5 above)
Comment (aramalipoor) [reply jenkins/jenkins.yml]: The whole story was assuming Github, for Gitlab a few other things must change as well, e.g. the Jenkinsfile.

How to do you think is best way to approach, should we add Gitlab stuff and comment them out? More or less this story gives CI developers a base ground to customize stuff.

For instance, a few things might not work with Gitlab as some plugins lack some features, e.g. https://github.com/jenkinsci/gitlab-plugin#pipeline-multibranch-jobs hooks are not triggered for Gitlab's Merge Requests (i.e. Pull Requests). So only works for branches, hence the Jenkinsfile needs to be tweaked a bit.

File: jenkins/jenkins.yml (hunk 6 above)
Comment (aramalipoor) [reply jenkins/jenkins.yml]: A new category `Continuous Integration & Deployment` already existing in UI but hidden as no templates belong there yet.

File: jenkins/jenkins.yml (hunk 3 above)
Comment (aramalipoor) [reply jenkins/jenkins.yml]: For OpenShift's pipeline to work the `Service` name must be `jenkins`. And as there's no point in having multiple Jenkins instances in the same project (I guess!) have done this on purpose. 

Jenkins already support multiple levels of authentication and distinguishing jobs and workspaces really well. 

WDYT?

File: jenkins/jenkins.yml (hunk 2 above)
Comment (aramalipoor) [reply jenkins/jenkins.yml]: Actually I'm not really sure what this does, I assumed it'll report the final generated URL so it'll output it as part of `new-app` or Add to Project, but with no luck. So I think i'll remove it for now.

File: jenkins/jenkins.yml (hunk 8 above)
Comment (aramalipoor) [reply jenkins/jenkins.yml]: Yeah ad it'll need more quickly when number of jobs increases. It's java anyways.

File: jenkins/jenkins.yml (hunk 1 above)
Comment (aramalipoor) [reply jenkins/jenkins.yml]: Yeah but I couldn't find a way to at least guess the URL since it'll include the project name. Is possible other than adding the route prefix parameter?

File: jenkins/jenkins.yml (hunk 7 above)
Comment (aramalipoor) [reply jenkins/jenkins.yml]: Done

File: jenkins/jenkins.yml (hunk 9 above)
Comment (aramalipoor) [reply jenkins/jenkins.yml]: Jenkins can't be scaled up like this. I have seen https://wiki.jenkins.io/display/JENKINS/Distributed+builds but didn't dive into it to see how it works.

File: jenkins/jenkins.yml (hunk 5 above)
Comment (alikhajeh1) [reply jenkins/jenkins.yml]: Plz leave a note on docs about this being only for GH and users should contact support to discuss requirements if they're using GL/BB
